### PR TITLE
APP-1086: Fire triggers once

### PR DESCRIPTION
**Pull request by davidnavas**
## What changes were proposed in this pull request?

Fire cleanups only once per type of thing to cleanup.

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)
(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

File: common/kvstore/src/main/java/org/apache/spark/util/kvstore/InMemoryStore.java [hunk 1]
```diff
@@ -201,16 +201,25 @@ public int size() {
...
+        Stream<T> stream = null;
+        if (first != null && last != null && asKey(first).compareTo(asKey(last)) == 0) {
+          // first is last, so we just want to filter the key, and don't care about sorting!
+          // This is the normal case for stage deletion -- there's no reason to sort a million
+          // summaries to pull one stage's keys on completion of a stage.
+          stream = sorted.stream().filter(e -> modifier * compare(e, getter, first) == 0);
+        } else {
+          LOG.info("Sorting InMemoryView of item count: " + sorted.size());
+          Collections.sort(sorted, (e1, e2) -> modifier * compare(e1, e2, getter));
+          stream = sorted.stream();
```
Comment (ianlcsd) [inline common/kvstore/src/main/java/org/apache/spark/util/kvstore/InMemoryStore.java]: do we need to sort at all, even first or last is not null?

Comment (davidnavas) [reply common/kvstore/src/main/java/org/apache/spark/util/kvstore/InMemoryStore.java]: for better or worse, that appears to be the contract.  If you can show that's excessive....

**Comment by davidnavas** [comment]
jttp

**Comment by davidnavas** [comment]
jttp

File: common/kvstore/src/main/java/org/apache/spark/util/kvstore/InMemoryStore.java (hunk 1 above)
Comment (davidnavas) [reply common/kvstore/src/main/java/org/apache/spark/util/kvstore/InMemoryStore.java]: fwiw: I suspect that the ordering is required for display purposes in the job page

**Review by markhamstra — APPROVED** [review]

**Pull request merged**